### ISSUE TITLE
Make formula editing on Android usable

### ISSUE THREAD

**Issue by orithena** [mkulesh/microMathematics#70]
Hi, I'm having some issues with formula editing on Android. None of the methods I tried allow me to add a simple "* cos( sqrt(x^2 + y^2) )" to the example formula of the 3D plot example.

Before I try to explain what I tried here in words, please see this [screencast](http://andaka.org/~dave/micromath.mp4).

I concede that this is my first attempt with this app, and overall, it looks great and would fill some of my use cases like no other app, but if I can't enter some basic formulae, it ceases to have any value :/

**Comment by orithena** [comment]
Well, a "formula editing mode" might be a good thing, e.g. when you tap a formula, it gets exclusive screen estate, is displayed bigger, edit mode stops with an OK button, etc...
Also, in this mode, a light gray button with a white + sign (for example) could be displayed at each possible insertion point to make it clear how the cursor could be moved to exactly that point.

I myself could also work with a LaTeX math syntax mode :)

**Comment by vi** [comment]
Can a formula be displayed as a tree for editing?

With tree operations like change operator, remove operator (bring operands up one level), introduce parentheses, swap two parts operands, duplicate operand and so on

Or can it just support editing as TeX?

**Comment by jinx100** [comment]
I like the formula entry and editing on Android. It is a new method but easy and forgiving after a little practice. This is a very polished math program and fun to use. Where can we share our work?

**Comment by mkulesh** [comment]
Hi @jinx100, if you like to share some your calculatons, please ether send me your files per e-mail (mikhail.kulesh@gmail.com) or just add these files to the diectory https://github.com/mkulesh/microMathematics/tree/master/app/src/main/assets/examples and open a pull reqest.

**Comment by lonetech** [comment]
The screencast demonstrates a few unintuitive UI issues (such as the zoom center), but most of it is spent on one central difficulty: placing the cursor for entering more text into a formula. This is largely because the formula entry has more of an active element than cursor position. When selecting somewhere in the formula, an entire element is selected, and when entering a function that function is applied to the element. The element being operated on in such a manner is highlighted in blue in the screencast. 

The workaround I found to achieve the desired operation of entering sqrt(x^2+y^2) is a bit tricky: First, I enter #x^2 for sqrt(x^2). Then, I long tap to select the x or 2. I then see that the blue selection doesn't cover the entire term, so I tap the widen button in the top toolbar; it's the one with four arrows, which might mean fullscreen, maximise or something in another program. Here, it widens the selection one operation, to x^2. Only then can I do the addition to add another term inside the square root. 

I also didn't find a way to make node selections with the keyboard.

**Comment by lonetech** [comment]
... what? Are you trying to argue that I should be browsing the ridiculously long toolbar strip with the touchscreen instead of directly writing on my very real keyboard (many devices have them, practically all can have them added)? And why, when this issue is in the first place about the difficulty in finding the function? How is your experience made any worse by making this function easier to find and use? The screencast posted in the original report shows specifically using the touchscreen navigation and **failing to find** the selection handling - which in turn perhaps shouldn't be so unidirectional (tap one element, then expand, no option to go from one element to e.g. one of its arguments). It's evident that the selection isn't obvious to everyone and you *didn't provide the guidance*, only kept repeating "it's fun, stop fighting the touchscreen". 

I also already described the entry process you only now described, as a Polish notation process. Back when the help was first asked for you posted "I like it", no instructions. This is a problem. That process works when you have the formula to enter in a mental AST (abstract syntax tree), but still doesn't solve the editing difficulty. Pointing out the selection and expand selection is still necessary for that. 

Your Android phone is not the only device out there. I started using this program on a Planet Computers Gemini PDA.

**Comment by jinx100** [comment]
You are correct that you were first to describe the selection and expansion process. I just tried to give a concise solution to the original complaint. 

I do suggest you use the "ridiculously" long strip, it is not that difficult. Android is a touch screen operating system and keyboards are optional.

**Comment by lonetech** [comment]
Except all the operations on that strip are the things I *can* easily type. It's the operations I can't do with it that are also hard or impossible with the keyboard. Like changing an addition to a subtraction (which in an actual formula is just changing the sign of a term). Or choosing where to place a new term, which requires this long-press + expand combination, that's awkward on touchscreen and inaccessible on keyboard. Just why does short pressing anywhere on a formula neither select anything nor place a cursor? Only the red boxes allow for easy entry.

**Comment by hubgitti** [comment]
Is it possible to have kind of a "plain text edit"-mode in your software, so that the data just can be entered or edited directly, without the use of these boxes?

**Comment by mkulesh** [comment]
Plus/Minus operators improved in the https://github.com/mkulesh/microMathematics/issues/97

**Comment by mkulesh** [comment]
Insertion before a selected term improved in https://github.com/mkulesh/microMathematics/issues/126
